src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -712,8 +715,32 @@ static int get_pool_offset(sd_dhcp_server *server, be32_t requested_ip) {
         return be32toh(requested_ip & ~server->netmask) - server->pool_offset;
 }
 
-static int dhcp_server_relay_message(sd_dhcp_server *server, DHCPMessage *message, size_t opt_length) {
+static int append_agent_information_option(sd_dhcp_server *server, DHCPMessage *message, size_t msg_len, size_t size) {
+        int r;
+        size_t offset;
+        uint8_t *end_option_ptr;
+
+        assert(server);
+        assert(message);
+
+        end_option_ptr = memrchr(message->options, SD_DHCP_OPTION_END, msg_len - sizeof(DHCPMessage));
```

Comment:
hmm, kan't SD_DHCP_OPTION_END show up in other options, too?

Reply:
but we are looking for it backward, end dhcp message must end with SD_DHCP_OPTION_END

Reply:
Maybe, it is safer to parse the message sent from upstream DHCP server from the beginning. The server may send broken message.

Reply:
Agree. Reworked option processing functions and extracted function to search for a particular option in options

Reply:
Ready for next review btw